doc: Add documentation for the new whitelist permissions

NicolasDorier opened the pull request:
Documenting the new feature https://github.com/bitcoin/bitcoin/pull/16248 . Ping Sjors .

Sjors commented:
@NicolasDorier can you include my "inbound" clarification from #16555 so I can close that?

Sjors commented on src/init.cpp:
```diff
@@ -438,8 +438,8 @@ void SetupServerArgs()
-    gArgs.AddArg("-whitebind=<addr>", "Bind to given address and whitelist peers connecting to it. Use [host]:port notation for IPv6", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
-    gArgs.AddArg("-whitelist=<IP address or network>", "Whitelist peers connecting from the given IP address (e.g. 1.2.3.4) or CIDR notated network (e.g. 1.2.3.0/24). Can be specified multiple times."
+    gArgs.AddArg("-whitebind=<[(permissions,)*@]addr>", "Bind to given address and whitelist peers connecting to it. Use [host]:port notation for IPv6.  Available permissions are: 'forcerelay,relay,noban,mempool,bloomfilter'. If the permission marker '@' is not specified, it will default to 'noban,mempool' with `whitelistrelay` and `whitelistforcerelay` optionally adding 'relay' and 'forcerelay'  ...
+    gArgs.AddArg("-whitelist=<[(permissions,)*@]IP address or network>", "Whitelist peers connecting from the given IP address (e.g. 1.2.3.4) or CIDR notated network (e.g. 1.2.3.0/24). Available permissions are: 'forcerelay,relay,noban,mempool,bloomfilter'. If the permission marker '@' is not specified, it will default to 'noban,mempool' with `whitelistrelay` and `whitelistforcerelay` optionally  ...
```
`whitelistrelay` is true by default, whereas `whitelistforcerelay` defaults to false, so a better wording would be:

```
If the permission marker '@' is not specified, it will default to 'noban,mempool,whitelistrelay', unless `-whitelistrelay=0`. If `-whitelistforcerelay=1` then 'forcerelay' is also set by default.
```

MarcoFalke replied:
Would it be possible to factor out the string with the list of permissions to a single symbol (mabye in net_permissions.h?)
Can you split the cpp string literal into muliple lines?

```cpp
   "safd1. "
   "foo2. "
   "foobar3. (default...)"

NicolasDorier replied:
Will do after #16555 get merged.

NicolasDorier commented:
@Sjors let's merge yours first.

NicolasDorier commented:
```

  -whitebind=<[(permissions,)*@]addr>
       Bind to given address and whitelist peers connecting to it. Use
       [host]:port notation for IPv6. If the permission marker '@' is
       specified, you can add a comma separated list of the following
       available permissions 'bloomfilter, noban, forcerelay, relay,
       mempool'. If the permission marker '@' is not specified, it will
       default to 'noban,mempool' with `whitelistrelay` and
       `whitelistforcerelay` optionally adding 'relay' and 'forcerelay'
       respectively. Can be specified multiple times.

  -whitelist=<[(permissions,)*@]IP address or network>
       Whitelist peers connecting from the given IP address (e.g. 1.2.3.4) or
       CIDR notated network (e.g. 1.2.3.0/24). If the permission marker
       '@' is specified, you can add a comma separated list of the
       following available permissions 'bloomfilter, noban, forcerelay,
       relay, mempool'. If the permission marker '@' is not specified,
       it will default to 'noban,mempool' with `whitelistrelay` and
       `whitelistforcerelay` optionally adding 'relay' and 'forcerelay'
       respectively. Can be specified multiple times.

  -whitelistforcerelay
       Force relay of transactions from whitelisted inbound peers even if the
       transactions were already in the mempool or violate local relay
       policy, this parameter is ignored if specific permissions are
       applied via whitebind/whitelist (default: 0)

  -whitelistrelay
       Accept relayed transactions received from whitelisted inbound peers even
       when not relaying transactions, this parameter is ignored if
       specific permissions are applied via whitebind/whitelist
       (default: 1)
```

NicolasDorier commented:
@MarcoFalke instead of using constant, I built in during runtime, it makes sure that the documentation is updated if the list of permission is updated.

MarcoFalke commented on src/init.cpp:
```diff
@@ -438,9 +438,16 @@ void SetupServerArgs()
+    for (const auto& permission : NetPermissions::ToStrings(PF_ALL)) {
+        permissionsDoc = strprintf("%s, %s", permissionsDoc, permission);
+    }
+    permissionsDoc = permissionsDoc.substr(2); // remove starting " ,"
```
This code should live in the net_permissions module

NicolasDorier replied:
Why? This is the only place it is needed.

harding commented:
@NicolasDorier I think this could be a bit more concise.  If possible, it'd also be good if it could briefly describe what the various options meant (even I had to look at the code to figure out what `mempool` provide, although BIP35 support was one of my guesses).  Here's a suggestion for white(bind|list), the other two looked ok to me:

```text
  -whitebind=<[(permissions,)*@]addr>
       Bind to given address and whitelist peers connecting to it. Use
       [host]:port notation for IPv6.  Allowed permissions are
       bloomfilter (allow requesting BIP37 filtered blocks and
       transactions), noban (do not ban for misbehavior), forcerelay
       (relay even non-standard transactions), relay (relay even in
       -blocksonly mode), and mempool (allow requesting BIP35 mempool
       contents).  Specify multiple permissions separated by commas
       (default: noban,mempool).  Can be specified multiple times

  -whitelist=<[(permissions,)*@]IP address or network>
       Whitelist peers connecting from the given IP address (e.g. 1.2.3.4) or
       CIDR notated network (e.g. 1.2.3.0/24).  Uses same permissions as
       -whitebind.  Can be specified multiple times
```

NicolasDorier commented:
@harding I like it, @MarcoFalke it means that we won't need your new string join util. :(
Will rebase tomorrow.

MarcoFalke commented:
> @MarcoFalke it means that we won't need your new string join util. :(

Good to hear. I like the new text as well. Will take a closer look in a couple of days.

NicolasDorier committed:
[Doc] Add documentation for the new whitelist permissions

NicolasDorier commented:
I just updated to use @harding suggestion, added documentation for bloomfilter permission as well.
I needed to change the doc for -whitelistforcerelay and -whitelistrelay as well.

NicolasDorier commented:
the linter says `whitebind` is undocumented?

harding commented:
@NicolasDorier bloomfilter was in my description so now you have it twice.  :-)

NicolasDorier commented:
oops, fixed it.

NicolasDorier commented:
Ok somehow this break the tests oO checking what happen.

Sjors commented:
ACK d536a10
<img width="550" alt="Schermafbeelding 2019-08-23 om 12 40 26" src="https://user-images.githubusercontent.com/10217/63587037-33bee380-c5a3-11e9-90bb-614009e06b4c.png">

<img width="549" alt="Schermafbeelding 2019-08-23 om 12 40 58" src="https://user-images.githubusercontent.com/10217/63587065-45a08680-c5a3-11e9-913c-0239915bf177.png">

Nit: remove spaces in examples `default: noban,mempool,relay` and `1.2.3.0/24`.

NicolasDorier commented:
Indeed, removed the spaces. Should be ready now.

NicolasDorier commented:
I broke the tests again. This is the hardest PR doc I did in my life.

NicolasDorier commented:
I rebased, now the error `rpc_invalidateblock` does not seem related to this PR.

MarcoFalke commented:
That is a known bug. See #16444. You don't need a passing travis on a doc-commit, it can be merged in any case after review.

NicolasDorier commented:
I think this one is ready to be merged @MarcoFalke

Sjors commented:
Indeed, re-ACK 66ad75472
